node/src/components/small_network/tasks.rs

```diff
@@ -0,0 +1,478 @@
+//! Tasks run by the component.
+
+use std::{
+    error::Error as StdError,
+    fmt::Display,
+    io,
+    net::SocketAddr,
+    pin::Pin,
+    sync::{Arc, Weak},
+    time::Duration,
+};
+
+use futures::{
+    future::{self, Either},
+    stream::{SplitSink, SplitStream},
+    Future, SinkExt, StreamExt,
+};
+use openssl::{
+    pkey::{PKey, Private},
+    ssl::Ssl,
+};
+use prometheus::IntGauge;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use tokio::{
+    net::TcpStream,
+    sync::{mpsc::UnboundedReceiver, watch},
+};
+use tokio_openssl::SslStream;
+use tracing::{
+    debug, error_span,
+    field::{self, Empty},
+    info, trace, warn, Instrument, Span,
+};
+
+use super::{
+    chain_info::ChainInfo,
+    counting_format::{ConnectionId, Role},
+    error::{display_error, ConnectionError, IoError},
+    event::{IncomingConnection, OutgoingConnection},
+    framed, Event, FramedTransport, Message, Payload, Transport,
+};
+use crate::{
+    components::networking_metrics::NetworkingMetrics,
+    reactor::{EventQueueHandle, QueueKind},
+    tls::{self, TlsCert},
+    types::NodeId,
+};
+
+// TODO: Constants need to be made configurable.
+
+/// Maximum time allowed to send or receive a handshake.
+const HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(20);
+
+/// Low-level TLS connection function.
+///
+/// Performs the actual TCP+TLS connection setup.
+async fn tls_connect<REv>(
+    context: &NetworkContext<REv>,
+    peer_addr: SocketAddr,
+) -> Result<(NodeId, Transport), ConnectionError>
+where
+    REv: 'static,
+{
+    let stream = TcpStream::connect(peer_addr)
+        .await
+        .map_err(ConnectionError::TcpConnection)?;
+
+    let mut transport = tls::create_tls_connector(&context.our_cert.as_x509(), &context.secret_key)
+        .and_then(|connector| connector.configure())
+        .and_then(|mut config| {
+            config.set_verify_hostname(false);
+            config.into_ssl("this-will-not-be-checked.example.com")
+        })
+        .and_then(|ssl| SslStream::new(ssl, stream))
+        .map_err(ConnectionError::TlsInitialization)?;
+
+    SslStream::connect(Pin::new(&mut transport))
+        .await
+        .map_err(ConnectionError::TlsHandshake)?;
+
+    let peer_cert = transport
+        .ssl()
+        .peer_certificate()
+        .ok_or(ConnectionError::NoPeerCertificate)?;
+
+    let peer_id = NodeId::from(
+        tls::validate_cert(peer_cert)
+            .map_err(ConnectionError::PeerCertificateInvalid)?
+            .public_key_fingerprint(),
+    );
+
```

Comment:
We used to check for the server having stopped at this point, and return an error rather than continue.  I seem to recall that this addressed an issue when transitioning from the joiner reactor to the validator.  Maybe I'm missing something, but it doesn't look like we're handling that case now in this PR?

Reply:
Correct me if I am wrong, but I believe you are referring to this? https://github.com/casper-network/casper-node/blob/2569ebd51923d0353542c93c5d47480246170cbb/node/src/components/small_network.rs#L1350-L1359

This was based on the following ticket "back then" (about 7 months ago):

> Problem: the joiner reactor connects to peers using small_network, and when the switch to the validator reactor occurs a new network is started and a new certificate is issued. As the the NodeId is based on the cert, this results in the same ip address having two different NodeId’s and two different ports being bound. Eventually the entry from the joiner reactor will go away when it is closed.
>
> Ideally, a clean finalization of the joiner reactor related state would clear any remaining incoming / outgoing connections before the validator reactor starts.

There are multiple reasons for removing this functionality, although not all aligned on a common goal:

1. The workaround was flawed: The code listed above turned a larger race condition into a smaller one. Core issue that an outgoing connection that is being established while the server is shut down will be kept open, since the handle that could cause a closing is stuck in the old event queue that is no longer being polled (the connection has a stream and a sink; the stream is closed as soon as possible, while the sink is shipped around with the `IncomingConnection` event). Adding the boolean flag would terminate it in many cases, but still leaves the door open for a connection to reach the event queue only to not be processed.

2. The original ticket is a non-issue. We should not be relying on the 1:1 identity of these IDs in the first place, I hope this has not been silently made a requirement, since we explicitly removed the persistence of IDs before that: https://github.com/casper-network/casper-node/commit/df7638a12d27170ee4b4d5e83ac9a94be5d5e7df#diff-4acfd86427b0a531d080292cd244708eaf7856fcd084459fce07ea3c6d2749ef

3. We silently re-introduced (2.) later on: https://github.com/casper-network/casper-node/commit/33dc7e5b4247e47021eb7fd6ab05c9bd293144cb#diff-5a2e839b41d4bdf6e8b228edbf0ef6b37bef11d7023cc469b06ee996301d8a76, making the original issue moot.

4. There is no longer potentially lingering state, as the connection process, which used to visit `SmallNetwork` three times during the process now only reports back with entirely finished connection. Otherwise, only the IP is known to the component. In a running component, these outgoing connections would also be cleaned up during housekeeping.

5. There was a proposed solution for cleaning up *everything* on shutdown as suggested in the ticket, but it was rejected: https://github.com/casper-network/casper-node/pull/1365

6. A situation where this is a problem will look like this: Node on machine A will have two outgoing connections, O->B and P->B, to the node on machine B. O->B will be from a dead reactor and a shutdown server tasks, so it will quickly be one-sided. There will be no messages outgoing from O->B as well, as there is no reactor, no sender for this (the handle is stuck in the unpolled event queue). P->B is from B's perspective an entirely different node, and any prior outgoing connection will have been terminated.
   O->B will just linger as dead weight, although we could clean it up, we currently don't purge incoming asymmetric connections, only outgoing ones.

If we are to prevent this in the future, a better way would be to simply nullify event queue handles once we've shutdown a reactor, and empty the queue entirely. Adding back the previous solution is a workaround at best though.

Reply:
OK - thanks for the detailed history and suggested future plan.  It does seem like this should be a non-issue, so I'll approve the PR as is.  I think we should be prepared to drop in a short-term workaround if required after further testing though.  We've been wrong about non-issues being non-issues before :)  Hopefully that won't be required though.

Reply:
Alright, permission to tell me "I told you so" officially granted!